extension/parquet/reader/variant_column_reader.cpp

```diff
@@ -0,0 +1,142 @@
+// Variant Column Reader
+//===--------------------------------------------------------------------===//
+VariantColumnReader::VariantColumnReader(ClientContext &context, ParquetReader &reader,
+                                         const ParquetColumnSchema &schema,
+                                         vector<unique_ptr<ColumnReader>> child_readers_p)
+    : ColumnReader(reader, schema), context(context), child_readers(std::move(child_readers_p)) {
+	D_ASSERT(Type().InternalType() == PhysicalType::VARCHAR);
+}
+
+ColumnReader &VariantColumnReader::GetChildReader(idx_t child_idx) {
+	if (!child_readers[child_idx]) {
+		throw InternalException("VariantColumnReader::GetChildReader(%d) - but this child reader is not set",
+		                        child_idx);
+	}
+	return *child_readers[child_idx].get();
+}
+
+void VariantColumnReader::InitializeRead(idx_t row_group_idx_p, const vector<ColumnChunk> &columns,
+                                         TProtocol &protocol_p) {
+	for (auto &child : child_readers) {
+		if (!child) {
+			continue;
+		}
+		child->InitializeRead(row_group_idx_p, columns, protocol_p);
+	}
+}
+
+static LogicalType GetIntermediateGroupType(optional_ptr<ColumnReader> typed_value) {
+	child_list_t<LogicalType> children;
+	children.emplace_back("value", LogicalType::BLOB);
+	if (typed_value) {
+		children.emplace_back("typed_value", typed_value->Type());
+	}
+	return LogicalType::STRUCT(std::move(children));
+}
+
+idx_t VariantColumnReader::Read(uint64_t num_values, data_ptr_t define_out, data_ptr_t repeat_out, Vector &result) {
+	if (pending_skips > 0) {
+		throw InternalException("VariantColumnReader cannot have pending skips");
+	}
+	optional_ptr<ColumnReader> typed_value_reader = child_readers.size() == 3 ? child_readers[2].get() : nullptr;
+
+	// If the child reader values are all valid, "define_out" may not be initialized at all
+	// So, we just initialize them to all be valid beforehand
+	std::fill_n(define_out, num_values, MaxDefine());
+
+	optional_idx read_count;
+
+	Vector metadata_intermediate(LogicalType::BLOB, num_values);
+	Vector intermediate_group(GetIntermediateGroupType(typed_value_reader), num_values);
+	auto &group_entries = StructVector::GetEntries(intermediate_group);
+	auto &value_intermediate = *group_entries[0];
+
+	auto metadata_values = child_readers[0]->Read(num_values, define_out, repeat_out, metadata_intermediate);
+	auto value_values = child_readers[1]->Read(num_values, define_out, repeat_out, value_intermediate);
+	if (metadata_values != value_values) {
+		throw InvalidInputException(
+		    "The Variant column did not contain the same amount of values for 'metadata' and 'value'");
+	}
+
+	auto result_data = FlatVector::GetData<string_t>(result);
+	auto &result_validity = FlatVector::Validity(result);
+
+	vector<VariantValue> conversion_result;
+	if (typed_value_reader) {
+		(void)typed_value_reader->Read(num_values, define_out, repeat_out, *group_entries[1]);
```

Comment:
Is there something we can verify here?
Should this logically always be equivalent to `value_values` and `metadata_values` ?